src/background_thread.c

```diff
@@ -0,0 +1,453 @@
+#define JEMALLOC_BACKGROUND_THREAD_C_
+#include "jemalloc/internal/jemalloc_preamble.h"
+#include "jemalloc/internal/jemalloc_internal_includes.h"
+
+#include "jemalloc/internal/assert.h"
+
+/******************************************************************************/
+/* Data. */
+
+/* This option should be opt-in only. */
+#define BACKGROUND_THREAD_DEFAULT false
+/* Read-only after initialization. */
+bool opt_background_thread = BACKGROUND_THREAD_DEFAULT;
+
+/* Protects global state during thread creation / termination. */
+malloc_mutex_t background_thread_lock;
+/* Indicates global state.  Atomic because decay reads this w/o locking. */
+atomic_b_t background_thread_enabled_state;
+size_t n_background_threads;
+/* Thread info per-index. */
+background_thread_info_t *background_thread_info;
+
+/******************************************************************************/
+
+#if !defined(JEMALLOC_BACKGROUND_THREAD)
+#define NOT_REACHED { not_reached(); }
+bool background_thread_create(tsd_t *tsd, unsigned arena_ind) NOT_REACHED
+bool background_threads_init(tsd_t *tsd) NOT_REACHED
+bool background_threads_enable(tsd_t *tsd) NOT_REACHED
+bool background_threads_disable(tsd_t *tsd) NOT_REACHED
+void background_thread_prefork(tsdn_t *tsdn) NOT_REACHED
+void background_thread_postfork_parent(tsdn_t *tsdn) NOT_REACHED
+void background_thread_postfork_child(tsdn_t *tsdn) NOT_REACHED
+#undef NOT_REACHED
+#else
+bool
+background_threads_init(tsd_t *tsd) {
+	assert(have_background_thread);
+	assert(narenas_total_get() > 0);
+
+	background_thread_enabled_set(tsd_tsdn(tsd), opt_background_thread);
+	if (malloc_mutex_init(&background_thread_lock,
+	    "background_thread_global",
+	    WITNESS_RANK_BACKGROUND_THREAD_GLOBAL)) {
+		return true;
+	}
+	background_thread_info = (background_thread_info_t *)base_alloc(
+	    tsd_tsdn(tsd), b0get(), ncpus * sizeof(background_thread_info_t),
+	    CACHELINE);
+	if (background_thread_info == NULL) {
+		return true;
+	}
+
+	for (unsigned i = 0 ; i < ncpus; i++) {
+		background_thread_info_t *info = &background_thread_info[i];
+		if (malloc_mutex_init(&info->mutex, "background_thread",
+		    WITNESS_RANK_BACKGROUND_THREAD)) {
+			return true;
+		}
+		if (pthread_cond_init(&info->cond, NULL)) {
+			return true;
+		}
+		info->started = false;
+	}
+
+	return false;
+}
+
+#define BILLION UINT64_C(1000000000)
+/* Threshold for determining when to wake up the background thread. */
+#define BACKGROUND_THREAD_NPAGES_THRESHOLD UINT64_C(1024)
+/* Background thread max sleep time is decay_time right shift by this. */
```

Comment:
I agree that with more dealloc activities, waking up every 100ms sounds very reasonable. However, if we do a fixed interval, on the less-active mode I'm still quite concerned about the wakeup frequency, given that multi-process would cause the wake up times to be multiplied by the number of processes.

I see the point that with the implementation right now we don't have a bound, i.e. we could decide to sleep long and then workload comes in. I think we can fix that part with the following design:
1) gets rid of the BACKGROUND_THREAD_MAX_INTERVAL. This makes the background thread could decide to sleep as long as the decay time.
2) in the interval search algorithm, consider only dirty pages in the backlog, i.e. remove the part that tries to "predict" future dirty pages based on average, in background_thread.c L118-L137.
3) This means background thread itself determines sleep interval solely based on backlog recorded dirty pages. To take future dirty pages into account, when we advance epoch we compute "new dirty pages need to be purged" since last background thread run -- this can be done accurately by walking through the last N epochs, or if we want to avoid the loop we can approximate by utilizing h_steps_sum (avg_new_pages_per_epoch * h_steps_sum[N]).

We do pay the complexity to gain the dynamic interval, but I feel the cost is worth it and makes the option more compelling. Being able to reduce the run frequency considerably also means reduce the interference to application, e.g. when p99 latency is of concern. With a fixed 100ms interval, assuming a two-process host we would still wake up >1k times a second even if no much work present.  With the background_thread option we move the purging or other expensive ops to async threads to avoid doing them on the application threads, so the goals include more predictable behavior and latency. I feel making the interval dynamic matches the motivation there.

Does the above design sound reasonable? If so I can try implementing it to evaluate further.